### PR TITLE
Introduce generic arith classes (on up-to-date master)

### PR DESCRIPTION
**Pull request by gogoex**
- Make classes used in range proofs to work with any set of initializer, point and scalar classes implementing required funcitons
- Make Elements class to work with any Scalar and Point classes
- Drop unused MulVec implementations

### CODE REVIEW

File: src/test/blsct/arith/mcl/mcl_g1point_tests.cpp [hunk 1]
```diff
@@ -262,95 +262,95 @@ BOOST_AUTO_TEST_CASE(test_g1point_hash_and_map)
...
 
-BOOST_AUTO_TEST_CASE(test_g1point_serialize_unserialize)
+BOOST_AUTO_TEST_CASE(test_serialize_unserialize)
 {
-    G1Point p(uint256::ONE);
+    MclG1Point p(uint256::ONE);
     CDataStream st(0, 0);
     p.Serialize(st);
-    BOOST_CHECK_EQUAL(st.size(), 49ul);
+    BOOST_CHECK_EQUAL(st.size(), 49ul); // TODO check why 49 instead of 48
```
Comment (aguycalled) [inline src/test/blsct/arith/mcl/mcl_g1point_tests.cpp:326]: vectors are serialised indicating the size first, then the vector content

**Comment by aguycalled** [comment]
would it be possible to add some documentation about what is the intended use for each class? Elements, Point, Points, MclG1Point, Scalar, Scalars, MclScalar, etc? ideally in the code for future reference

File: src/test/blsct/arith/mcl/mcl_g1point_tests.cpp (hunk 1 above)
Comment (gogoex) [reply src/test/blsct/arith/mcl/mcl_g1point_tests.cpp:326]: Ah i see. Thank you.

**Comment by gogoex** [comment]
Sure. I will add a mark down for it.

**Comment by aguycalled** [comment]
liking how you even documented the different approaches you considered 👍

**Comment by gogoex** [comment]
Great. Can this PR be merged now? Regarding the remaining 2 failing checks, one is `Intel macOS instances are no longer supported!` and the other is `LINK : fatal error LNK1104: cannot open file 'bls384_256.lib'` which I don't think I will be able to fix.

**Comment by aguycalled** [comment]
> Great. Can this PR be merged now? Regarding the remaining 2 failing checks, one is `Intel macOS instances are no longer supported!` and the other is `LINK : fatal error LNK1104: cannot open file 'bls384_256.lib'` which I don't think I will be able to fix.

I've created a PR to fix the macOS task: https://github.com/navcoin/navcoin/pull/88/files
For fixing the LNK1104 issue I'd need a windows computer, which I won't have access until the end of the month or so. It's not urgent anyway.

I want to have another extra look through the PR and then I will approve and merge 👍

**Review by aguycalled — APPROVED** [review]

**Pull request merged**